InCodeGenerator fails for LongDecimalType

```sql
select * from tpch.tiny.orders where rand() < 0 or cast(orderkey as decimal(38)) in (1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33);
```

```
io.prestosql.spi.PrestoException: Compiler failed
	at io.prestosql.sql.planner.LocalExecutionPlanner$Visitor.visitScanFilterAndProject(LocalExecutionPlanner.java:1303)
	at io.prestosql.sql.planner.LocalExecutionPlanner$Visitor.visitFilter(LocalExecutionPlanner.java:1163)
	at io.prestosql.sql.planner.LocalExecutionPlanner$Visitor.visitFilter(LocalExecutionPlanner.java:705)
	at io.prestosql.sql.planner.plan.FilterNode.accept(FilterNode.java:72)
	at io.prestosql.sql.planner.LocalExecutionPlanner.plan(LocalExecutionPlanner.java:461)
	at io.prestosql.sql.planner.LocalExecutionPlanner.plan(LocalExecutionPlanner.java:383)
	at io.prestosql.execution.SqlTaskExecutionFactory.create(SqlTaskExecutionFactory.java:75)
	at io.prestosql.execution.SqlTask.updateTask(SqlTask.java:382)
	at io.prestosql.execution.SqlTaskManager.updateTask(SqlTaskManager.java:383)
	at io.prestosql.server.TaskResource.createOrUpdateTask(TaskResource.java:129)
	at sun.reflect.GeneratedMethodAccessor900.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory.lambda$static$0(ResourceMethodInvocationHandlerFactory.java:76)
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:148)
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:191)
	at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$ResponseOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:200)
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:103)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:493)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:415)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:104)
	at org.glassfish.jersey.server.ServerRuntime$1.run(ServerRuntime.java:277)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:272)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:268)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:316)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:298)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:268)
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:289)
	at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:256)
	at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:703)
	at org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:416)
	at org.glassfish.jersey.servlet.WebComponent.service(WebComponent.java:370)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:389)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:342)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:229)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:755)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1617)
	at io.prestosql.server.security.AuthenticationFilter.withAuthenticatedIdentity(AuthenticationFilter.java:207)
	at io.prestosql.server.security.AuthenticationFilter.doFilter(AuthenticationFilter.java:99)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1604)
	at io.airlift.http.server.TraceTokenFilter.doFilter(TraceTokenFilter.java:63)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1604)
	at io.airlift.http.server.TimingFilter.doFilter(TimingFilter.java:51)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1604)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:545)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
	at org.eclipse.jetty.server.handler.gzip.GzipHandler.handle(GzipHandler.java:717)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:235)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1300)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:485)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1215)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:146)
	at org.eclipse.jetty.server.handler.StatisticsHandler.handle(StatisticsHandler.java:173)
	at org.eclipse.jetty.server.handler.HandlerList.handle(HandlerList.java:59)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
	at org.eclipse.jetty.server.Server.handle(Server.java:500)
	at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:383)
	at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:547)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:375)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:273)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103)
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:117)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:336)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:313)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:171)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:129)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:375)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:806)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:938)
	at java.lang.Thread.run(Thread.java:748)
Caused by: com.google.common.util.concurrent.UncheckedExecutionException: io.prestosql.spi.PrestoException: GENERIC_INTERNAL_ERROR
	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2050)
	at com.google.common.cache.LocalCache.get(LocalCache.java:3952)
	at com.google.common.cache.LocalCache.getOrLoad(LocalCache.java:3974)
	at com.google.common.cache.LocalCache$LocalLoadingCache.get(LocalCache.java:4958)
	at com.google.common.cache.LocalCache$LocalLoadingCache.getUnchecked(LocalCache.java:4964)
	at io.prestosql.sql.gen.ExpressionCompiler.compileCursorProcessor(ExpressionCompiler.java:82)
	at io.prestosql.sql.planner.LocalExecutionPlanner$Visitor.visitScanFilterAndProject(LocalExecutionPlanner.java:1266)
	... 74 more
Caused by: io.prestosql.spi.PrestoException: GENERIC_INTERNAL_ERROR
	at io.prestosql.util.FastutilSetHelper$ObjectStrategy.equals(FastutilSetHelper.java:214)
	at it.unimi.dsi.fastutil.objects.ObjectOpenCustomHashSet.add(ObjectOpenCustomHashSet.java:302)
	at java.util.AbstractCollection.addAll(AbstractCollection.java:344)
	at it.unimi.dsi.fastutil.objects.ObjectOpenCustomHashSet.addAll(ObjectOpenCustomHashSet.java:297)
	at it.unimi.dsi.fastutil.objects.ObjectOpenCustomHashSet.<init>(ObjectOpenCustomHashSet.java:137)
	at io.prestosql.util.FastutilSetHelper.toFastutilHashSet(FastutilSetHelper.java:62)
	at io.prestosql.sql.gen.InCodeGenerator.generateExpression(InCodeGenerator.java:223)
	at io.prestosql.sql.gen.RowExpressionCompiler$Visitor.visitSpecialForm(RowExpressionCompiler.java:155)
	at io.prestosql.sql.gen.RowExpressionCompiler$Visitor.visitSpecialForm(RowExpressionCompiler.java:80)
	at io.prestosql.sql.relational.SpecialForm.accept(SpecialForm.java:90)
	at io.prestosql.sql.gen.RowExpressionCompiler.compile(RowExpressionCompiler.java:77)
	at io.prestosql.sql.gen.BytecodeGeneratorContext.generate(BytecodeGeneratorContext.java:77)
	at io.prestosql.sql.gen.BytecodeGeneratorContext.generate(BytecodeGeneratorContext.java:72)
	at io.prestosql.sql.gen.OrCodeGenerator.generateExpression(OrCodeGenerator.java:44)
	at io.prestosql.sql.gen.RowExpressionCompiler$Visitor.visitSpecialForm(RowExpressionCompiler.java:155)
	at io.prestosql.sql.gen.RowExpressionCompiler$Visitor.visitSpecialForm(RowExpressionCompiler.java:80)
	at io.prestosql.sql.relational.SpecialForm.accept(SpecialForm.java:90)
	at io.prestosql.sql.gen.RowExpressionCompiler.compile(RowExpressionCompiler.java:77)
	at io.prestosql.sql.gen.RowExpressionCompiler.compile(RowExpressionCompiler.java:72)
	at io.prestosql.sql.gen.CursorProcessorCompiler.generateFilterMethod(CursorProcessorCompiler.java:251)
	at io.prestosql.sql.gen.CursorProcessorCompiler.generateMethods(CursorProcessorCompiler.java:82)
	at io.prestosql.sql.gen.ExpressionCompiler.compileProcessor(ExpressionCompiler.java:153)
	at io.prestosql.sql.gen.ExpressionCompiler.compile(ExpressionCompiler.java:133)
	at io.prestosql.sql.gen.ExpressionCompiler.lambda$new$0(ExpressionCompiler.java:69)
	at com.google.common.cache.CacheLoader$FunctionToCacheLoader.load(CacheLoader.java:165)
	at com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3528)
	at com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2277)
	at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2154)
	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2044)
	... 80 more
Caused by: java.lang.NullPointerException
	at io.prestosql.spi.type.UnscaledDecimal128Arithmetic.getRawInt(UnscaledDecimal128Arithmetic.java:1725)
	at io.prestosql.spi.type.UnscaledDecimal128Arithmetic.isNegative(UnscaledDecimal128Arithmetic.java:750)
	at io.prestosql.spi.type.UnscaledDecimal128Arithmetic.isStrictlyNegative(UnscaledDecimal128Arithmetic.java:740)
	at io.prestosql.spi.type.UnscaledDecimal128Arithmetic.compare(UnscaledDecimal128Arithmetic.java:668)
	at io.prestosql.type.DecimalInequalityOperators.boxedLongLong(DecimalInequalityOperators.java:155)
	at io.prestosql.util.FastutilSetHelper$ObjectStrategy.equals(FastutilSetHelper.java:206)
	... 108 more
```

--- a/presto-main/src/main/java/io/prestosql/util/FastutilSetHelper.java
+++ b/presto-main/src/main/java/io/prestosql/util/FastutilSetHelper.java
@@ -188,6 +188,9 @@ public final class FastutilSetHelper
         @Override
         public int hashCode(Object value)
         {
+            if (value == null) {
+                return 0;
+            }
             try {
                 return Long.hashCode((long) hashCodeHandle.invokeExact(value));
             }
@@ -201,6 +204,9 @@ public final class FastutilSetHelper
         @Override
         public boolean equals(Object a, Object b)
         {
+            if (b == null || a == null) {
+                return a == null && b == null;
+            }
             try {
                 Boolean result = (Boolean) equalsHandle.invokeExact(a, b);
                 // FastutilHashSet is not intended be used for indeterminate values lookup

--- a/presto-main/src/test/java/io/prestosql/operator/scalar/FunctionAssertions.java
+++ b/presto-main/src/test/java/io/prestosql/operator/scalar/FunctionAssertions.java
@@ -53,6 +53,7 @@ import io.prestosql.spi.connector.RecordPageSource;
 import io.prestosql.spi.connector.RecordSet;
 import io.prestosql.spi.predicate.TupleDomain;
 import io.prestosql.spi.predicate.Utils;
+import io.prestosql.spi.type.DecimalType;
 import io.prestosql.spi.type.RowType;
 import io.prestosql.spi.type.TimeZoneKey;
 import io.prestosql.spi.type.Type;
@@ -77,6 +78,7 @@ import org.openjdk.jol.info.ClassLayout;
 import java.io.Closeable;
 import java.lang.reflect.Field;
 import java.lang.reflect.Modifier;
+import java.math.BigDecimal;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collections;
@@ -97,8 +99,10 @@ import static io.prestosql.SessionTestUtils.TEST_SESSION;
 import static io.prestosql.block.BlockAssertions.createBooleansBlock;
 import static io.prestosql.block.BlockAssertions.createDoublesBlock;
 import static io.prestosql.block.BlockAssertions.createIntsBlock;
+import static io.prestosql.block.BlockAssertions.createLongDecimalsBlock;
 import static io.prestosql.block.BlockAssertions.createLongsBlock;
 import static io.prestosql.block.BlockAssertions.createRowBlock;
+import static io.prestosql.block.BlockAssertions.createShortDecimalsBlock;
 import static io.prestosql.block.BlockAssertions.createSlicesBlock;
 import static io.prestosql.block.BlockAssertions.createStringsBlock;
 import static io.prestosql.block.BlockAssertions.createTimestampsWithTimezoneBlock;
@@ -109,6 +113,8 @@ import static io.prestosql.spi.StandardErrorCode.NUMERIC_VALUE_OUT_OF_RANGE;
 import static io.prestosql.spi.type.BigintType.BIGINT;
 import static io.prestosql.spi.type.BooleanType.BOOLEAN;
 import static io.prestosql.spi.type.DateTimeEncoding.packDateTimeWithZone;
+import static io.prestosql.spi.type.DecimalType.createDecimalType;
+import static io.prestosql.spi.type.Decimals.encodeScaledValue;
 import static io.prestosql.spi.type.DoubleType.DOUBLE;
 import static io.prestosql.spi.type.IntegerType.INTEGER;
 import static io.prestosql.spi.type.TimestampWithTimeZoneType.TIMESTAMP_WITH_TIME_ZONE;
@@ -142,6 +148,8 @@ public final class FunctionAssertions
     private static final int TEST_ROW_NUMBER_OF_FIELDS = 2500;
     private static final RowType TEST_ROW_TYPE = createTestRowType(TEST_ROW_NUMBER_OF_FIELDS);
     private static final Block TEST_ROW_DATA = createTestRowData(TEST_ROW_TYPE);
+    private static final DecimalType SHORT_DECIMAL_TYPE = createDecimalType(14);
+    private static final DecimalType LONG_DECIMAL_TYPE = createDecimalType(28);
 
     private static final Page SOURCE_PAGE = new Page(
             createLongsBlock(1234L),
@@ -154,7 +162,9 @@ public final class FunctionAssertions
             createTimestampsWithTimezoneBlock(packDateTimeWithZone(new DateTime(1970, 1, 1, 0, 1, 0, 999, DateTimeZone.UTC).getMillis(), TimeZoneKey.getTimeZoneKey("Z"))),
             createSlicesBlock(Slices.wrappedBuffer((byte) 0xab)),
             createIntsBlock(1234),
-            TEST_ROW_DATA);
+            TEST_ROW_DATA,
+            createShortDecimalsBlock("1234"),
+            createLongDecimalsBlock("1234"));
 
     private static final Page ZERO_CHANNEL_PAGE = new Page(1);
 
@@ -170,6 +180,8 @@ public final class FunctionAssertions
             .put(new Symbol("bound_binary_literal"), VARBINARY)
             .put(new Symbol("bound_integer"), INTEGER)
             .put(new Symbol("bound_row"), TEST_ROW_TYPE)
+            .put(new Symbol("bound_short_decimal"), SHORT_DECIMAL_TYPE)
+            .put(new Symbol("bound_long_decimal"), LONG_DECIMAL_TYPE)
             .build());
 
     private static final Map<Symbol, Integer> INPUT_MAPPING = ImmutableMap.<Symbol, Integer>builder()
@@ -184,6 +196,8 @@ public final class FunctionAssertions
             .put(new Symbol("bound_binary_literal"), 8)
             .put(new Symbol("bound_integer"), 9)
             .put(new Symbol("bound_row"), 10)
+            .put(new Symbol("bound_short_decimal"), 11)
+            .put(new Symbol("bound_long_decimal"), 12)
             .build();
 
     private static final PageSourceProvider PAGE_SOURCE_PROVIDER = new TestPageSourceProvider();
@@ -850,7 +864,7 @@ public final class FunctionAssertions
             assertInstanceOf(split.getConnectorSplit(), FunctionAssertions.TestSplit.class);
             FunctionAssertions.TestSplit testSplit = (FunctionAssertions.TestSplit) split.getConnectorSplit();
             if (testSplit.isRecordSet()) {
-                RecordSet records = InMemoryRecordSet.builder(ImmutableList.of(BIGINT, VARCHAR, DOUBLE, BOOLEAN, BIGINT, VARCHAR, VARCHAR, TIMESTAMP_WITH_TIME_ZONE, VARBINARY, INTEGER, TEST_ROW_TYPE))
+                RecordSet records = InMemoryRecordSet.builder(ImmutableList.of(BIGINT, VARCHAR, DOUBLE, BOOLEAN, BIGINT, VARCHAR, VARCHAR, TIMESTAMP_WITH_TIME_ZONE, VARBINARY, INTEGER, TEST_ROW_TYPE, SHORT_DECIMAL_TYPE, LONG_DECIMAL_TYPE))
                         .addRow(
                                 1234L,
                                 "hello",
@@ -862,7 +876,9 @@ public final class FunctionAssertions
                                 packDateTimeWithZone(new DateTime(1970, 1, 1, 0, 1, 0, 999, DateTimeZone.UTC).getMillis(), TimeZoneKey.getTimeZoneKey("Z")),
                                 Slices.wrappedBuffer((byte) 0xab),
                                 1234,
-                                TEST_ROW_DATA.getObject(0, Block.class))
+                                TEST_ROW_DATA.getObject(0, Block.class),
+                                new BigDecimal("1234").unscaledValue().longValue(),
+                                encodeScaledValue(new BigDecimal("1234")))
                         .build();
                 return new RecordPageSource(records);
             }

--- a/presto-main/src/test/java/io/prestosql/sql/gen/TestExpressionCompiler.java
+++ b/presto-main/src/test/java/io/prestosql/sql/gen/TestExpressionCompiler.java
@@ -1321,6 +1321,18 @@ public class TestExpressionCompiler
         assertExecute("bound_timestamp_with_timezone in (" + timestampValues + ")", BOOLEAN, true);
         assertExecute("bound_timestamp_with_timezone in (TIMESTAMP '1970-01-01 01:01:00.0+02:00')", BOOLEAN, false);
 
+        String shortDecimalValues = range(2000, 7000)
+                .mapToObj(value -> format("decimal '%s'", value))
+                .collect(joining(", "));
+        assertExecute("bound_short_decimal in (1234, " + shortDecimalValues + ")", BOOLEAN, true);
+        assertExecute("bound_short_decimal in (" + shortDecimalValues + ")", BOOLEAN, false);
+
+        String longDecimalValues = range(2000, 7000)
+                .mapToObj(value -> format("decimal '123456789012345678901234567890%s'", value))
+                .collect(joining(", "));
+        assertExecute("bound_long_decimal in (1234, " + longDecimalValues + ")", BOOLEAN, true);
+        assertExecute("bound_long_decimal in (" + longDecimalValues + ")", BOOLEAN, false);
+
         Futures.allAsList(futures).get();
     }
 

--- a/presto-spi/src/main/java/io/prestosql/spi/connector/InMemoryRecordSet.java
+++ b/presto-spi/src/main/java/io/prestosql/spi/connector/InMemoryRecordSet.java
@@ -30,6 +30,8 @@ import java.util.List;
 import static io.prestosql.spi.type.BigintType.BIGINT;
 import static io.prestosql.spi.type.BooleanType.BOOLEAN;
 import static io.prestosql.spi.type.DateType.DATE;
+import static io.prestosql.spi.type.Decimals.isLongDecimal;
+import static io.prestosql.spi.type.Decimals.isShortDecimal;
 import static io.prestosql.spi.type.DoubleType.DOUBLE;
 import static io.prestosql.spi.type.IntegerType.INTEGER;
 import static io.prestosql.spi.type.TimestampType.TIMESTAMP;
@@ -242,6 +244,14 @@ public class InMemoryRecordSet
                 else if (type instanceof RowType) {
                     checkArgument(value instanceof Block,
                             "Expected value %d to be an instance of Block, but is a %s", i, value.getClass().getSimpleName());
+                }
+                else if (isShortDecimal(type)) {
+                    checkArgument(value instanceof Long,
+                            "Expected value %d to be an instance of Long, but is a %s", i, value.getClass().getSimpleName());
+                }
+                else if (isLongDecimal(type)) {
+                    checkArgument(value instanceof Slice,
+                            "Expected value %d to be an instance of Slice, but is a %s", i, value.getClass().getSimpleName());
                 }
                 else {
                     throw new IllegalStateException("Unsupported column type " + types.get(i));

